miden-rpc/src/server/api.rs

```diff
@@ -0,0 +1,122 @@
+use std::{net::ToSocketAddrs, sync::Arc};
+use tracing::info;
+use anyhow::Result;
+use crate::config::RPCConfig;
+use tonic::{transport::{Channel, Server, Error}, Request, Response, Status, Streaming};
+use prost::Message;
+use miden_proto::{store::{self, api_client}, rpc, digest};
+
+pub struct RpcApi {
+    client: Arc<api_client::ApiClient<Channel>>,
+}
+
+impl RpcApi {
+    async fn from_config(config: &RPCConfig) -> Result<Self, Error> {
+        let client = api_client::ApiClient::connect(config.db.clone()).await?;
+        info!(
+            db = config.db,
+            "DB client initialized",
+        );
+        Ok(Self {
+            client: Arc::new(client),
+        })
+    }
+}
+
+#[tonic::async_trait]
+impl rpc::api_server::Api for RpcApi {
+    type CheckNullifiersStream = Streaming<rpc::ResponseBytes>;
+    type ListNotesByTagStream = Streaming<rpc::ResponseBytes>;
+
+    async fn get_block_header(
+        &self,
+        request: Request<rpc::GetBlockHeaderRequest>,
+    ) -> Result<Response<rpc::ResponseBytes>, Status>
+    {
+        let mut client = self.client.as_ref().clone();
+        let inner = request.into_inner();
+        let request = store::GetBlockHeaderRequest{block_spec: inner.block_spec};
+        let response = client.get_block_header(request).await?;
+        let data = response.into_inner().encode_to_vec();
+        Ok(Response::new(rpc::ResponseBytes{data}))
+    }
+
+    async fn get_note(
+        &self,
+        request: Request<rpc::GetNoteRequest>,
+    ) -> Result<Response<rpc::ResponseBytes>, Status>
+    {
+        let mut client = self.client.as_ref().clone();
+        let inner = request.into_inner();
+        let request = store::GetNoteRequest{note_spec: inner.note_spec};
+        let response = client.get_note(request).await?;
+        let data = response.into_inner().encode_to_vec();
+        Ok(Response::new(rpc::ResponseBytes{data}))
+    }
```

Comment:
Not for this PR but just a general question: is there a way to just "re-stream" the data so that we don't have to allocate a vector for it?

Reply:
There are a few issues with that:

- The connections support compression, it is unlikely just pushing compressed data from stream to another will work as intended
- The generated client uses the streaming api, which automatically decodes the data. So there is no easy way of getting the underlying buffer by using that abstraction.
  -  Circumventing this means that we have to write our own gRPC client, handling http2, buffering, compression, which loses all the benefits of tonic/prost/tower
- the type `ResponseBytes` is not the same as the type of the response, so the encoding differs and the message needs to be re-encoded

Reply:
Makes sense. I think what we have is fine for now and we can look into it more if/when it becomes a problem.